drivers/modem/ublox-sara-n310.c

```diff
@@ -835,6 +897,148 @@ static bool offload_is_supported(int family, int type, int proto)
 NET_SOCKET_REGISTER(ublox_sara_n310, AF_UNSPEC, offload_is_supported,
 		    offload_socket);
 
+/*
+ * Pin functions.
+ *
+ * The SARA-N310 forces the power pin low to achieve the following:
+ * min 1s - max 2.5s: trigger module switch-on
+ * min 1s - max 2.5s: trigger module wake-up from PSM
+ * min 2.5s: trigger module switch-off
+ */
+static int turn_on_module(void)
+{
+	int ret, timeout = 0;
+
+	LOG_DBG("MDM_POWER_PIN -> ENABLE");
+	modem_pin_write(&mdata.context, MDM_POWER, MDM_POWER_ENABLE);
+	k_sleep(K_MSEC(1500));
+
+	LOG_DBG("MDM_POWER_PIN -> DISABLE");
+	modem_pin_write(&mdata.context, MDM_POWER, MDM_POWER_DISABLE);
+
+	while (!is_awake() && (timeout++ < 5)) {
+		k_sleep(K_MSEC(10));
+	}
+
+	if (timeout >= 5) {
+		ret = -timeout;
+
+		return ret;
+	}
+
+	return 0;
+}
+
+static int turn_off_module(void)
+{
+	LOG_DBG("MDM_POWER_PIN -> ENABLE");
+	modem_pin_write(&mdata.context, MDM_POWER, MDM_POWER_ENABLE);
+	k_sleep(K_SECONDS(3));
+
+	LOG_DBG("MDM_POWER_PIN -> DISABLE");
+	modem_pin_write(&mdata.context, MDM_POWER, MDM_POWER_DISABLE);
+
+	return 0;
+}
+
+/* check if module is awake by checking V_INT value */
+static int is_awake(void)
+{
+	return modem_pin_read(&mdata.context, MDM_VINT);
+}
+
+/* PSM functions, exposed to application through header */
+/* Power mode setting, +NVSETPM */
+int n310_psm_set_mode(int psm_mode)
```

Comment:
[Minor and optional] Would be nice to have an enum with the allowed states. In this way you can help the user of the function